### PR TITLE
Avoid byte-extraction on arrays 

### LINKED ISSUE #169: Segmentation fault while encoding in Boolector
Executing esbmc with boolector and the flags `--no-bounds-check --no-pointer-check --no-div-by-zero-check --incremental-bmc --max-k-step 10` gives a Segmentation fault in the following program:

```
a;
void h();
#define c(d, g) strcmp(d, 0)
main(argc, argv) char **argv;
{
  char *b[argc];
  char e;
  for (; e;)
    *b = a;
  h(b);
}
void h(f) { c(f, ); }
```

Using Z3 this fault does not happen.

### PR DESCRIPTION
**Pull request by fbrausse**
This PR
- fixes an out-of-bounds vector access in the non-byte array case in dereferencet::extract_bytes_from_array(),
- refactors extract_bytes_*() set of functions to a more unified state with fewer case distinctions,
- strengthens the conditions determining `overflows_boundaries` for extracting elements from arrays, and
- removes an unnecessary bitcast-to-BV before byte-extraction.

This results in byte-extraction directly on symbolically sized arrays being avoided. We need to avoid it there because smt_convt::convert_byte_extract() can't deal with those (it expects type->get_width() to work).

 Fixes #169.

About refining the definition of `overflows_boundaries`, which is used in the condition on when simple recursion is allowed in construct_from_array(): just checking whether `deref_size <= subtype_size` is not enough to guarantee that we can just index into the array; need to take alignment and the offset into account as well in order to prove that we access only a single element of the array.

We should run this over SV-COMP as well, in particular due to the points described in the last paragraph.

Some `--fixedbv` and some `--ir` regressions start to fail with this PR, the latter mainly due to
```
ERROR: Refusing to byte extract in integer mode; re-run in bitvector mode
```

I wonder what significance the `--fixedbv` regression tests have, given that there hardly are any real-world C programs that assume `float` means fixed-point 16.16.

### CODE REVIEW

**Comment by fbrausse** [comment]
> Some --fixedbv and some --ir regressions start to fail with this PR

These failing tests can be fixed by removing the pessimization on `alignment`:
```diff
diff --git a/src/pointer-analysis/dereference.cpp b/src/pointer-analysis/dereference.cpp
index c3ad805bc5..7baebf5015 100644
--- a/src/pointer-analysis/dereference.cpp
+++ b/src/pointer-analysis/dereference.cpp
@@ -695,13 +695,6 @@ expr2tc dereferencet::build_reference_to(
   {
     assert(alignment != 0);
 
-    if(!is_symbol2t(deref_expr))
-    {
-      // The expression being dereferenced isn't just a symbol: it might have
-      // all kind of things messing with alignment in there. We could interpret
-      // it as future work.
-      alignment = 1;
-    }
     final_offset =
       pointer_offset2tc(get_int_type(config.ansi_c.address_width), deref_expr);
   }
```
Though I'm not confident yet that this would be sound.

File: src/pointer-analysis/dereference.cpp [hunk 1]
```diff
@@ -1911,36 +1938,24 @@ std::vector<expr2tc> dereferencet::extract_bytes_from_array(
...
+
   // Calculating the array index based on the given byte offset
-  expr2tc accuml_offs =
+  expr2tc elem_idx =
     div2tc(offset->type, offset, gen_long(offset->type, bytes_per_index));
-  for(unsigned int i = 0; i < num_bytes; i++)
+  expr2tc zero = gen_zero(offset->type);
+  expr2tc elem_offs =
+    modulus2tc(offset->type, offset, gen_long(offset->type, bytes_per_index));
+  for(unsigned i = 0, n; i < num_bytes; i += n)
```
Comment (shmarovfedor) [inline src/pointer-analysis/dereference.cpp]: Could you add a comment that gives more details (perhaps one-two examples) on what this loop does?

**Comment by lucasccordeiro** [comment]
Hi @fbrausse,

We did much research in the past to verify and synthesize digital controllers and filters using fixed-point arithmetic, e.g., [1-5].

Also, we did much research in the past to optimize mathematical functions using integers and reals from the mathematical domain, e.g., [6-9].

So, it would be great to keep these options working on ESBMC. Also, CERN uses integers and reals to verify the Programmable Logic Controller for Particle Acceleration.

[1] [Alessandro Abate](https://dblp.uni-trier.de/pid/19/3904.html), [Iury Bessa](https://dblp.uni-trier.de/pid/143/7442.html), Lucas C. Cordeiro, [Cristina David](https://dblp.uni-trier.de/pid/01/3840.html), [Pascal Kesseli](https://dblp.uni-trier.de/pid/168/8498.html), [Daniel Kroening](https://dblp.uni-trier.de/pid/k/DanielKroening.html), [Elizabeth Polgreen](https://dblp.uni-trier.de/pid/183/7353.html):
Automated formal synthesis of provably safe digital controllers for continuous plants. [Acta Informatica 57(1-2)](https://dblp.uni-trier.de/db/journals/acta/acta57.html#AbateBCDKKP20): 223-244 (2020)
[2] [Alessandro Abate](https://dblp.uni-trier.de/pid/19/3904.html), [Iury Bessa](https://dblp.uni-trier.de/pid/143/7442.html), [Dario Cattaruzza](https://dblp.uni-trier.de/pid/164/5620.html), Lucas C. Cordeiro, [Cristina David](https://dblp.uni-trier.de/pid/01/3840.html), [Pascal Kesseli](https://dblp.uni-trier.de/pid/168/8498.html), [Daniel Kroening](https://dblp.uni-trier.de/pid/k/DanielKroening.html), [Elizabeth Polgreen](https://dblp.uni-trier.de/pid/183/7353.html):
Automated Formal Synthesis of Digital Controllers for State-Space Physical Plants. [CAV (1) 2017](https://dblp.uni-trier.de/db/conf/cav/cav2017-1.html#AbateBCCDKKP17): 462-482
[3] [Alessandro Abate](https://dblp.uni-trier.de/pid/19/3904.html), [Iury Bessa](https://dblp.uni-trier.de/pid/143/7442.html), [Dario Cattaruzza](https://dblp.uni-trier.de/pid/164/5620.html), Lucas C. Cordeiro, [Cristina David](https://dblp.uni-trier.de/pid/01/3840.html), [Pascal Kesseli](https://dblp.uni-trier.de/pid/168/8498.html), [Daniel Kroening](https://dblp.uni-trier.de/pid/k/DanielKroening.html):
Sound and Automated Synthesis of Digital Stabilizing Controllers for Continuous Plants. [HSCC 2017](https://dblp.uni-trier.de/db/conf/hybrid/hscc2017.html#AbateBCCDKK17): 197-206
[4] [Alessandro Abate](https://dblp.uni-trier.de/pid/19/3904.html), [Iury Bessa](https://dblp.uni-trier.de/pid/143/7442.html), [Dario Cattaruzza](https://dblp.uni-trier.de/pid/164/5620.html), [Lennon C. Chaves](https://dblp.uni-trier.de/pid/188/5666.html), Lucas C. Cordeiro, [Cristina David](https://dblp.uni-trier.de/pid/01/3840.html), [Pascal Kesseli](https://dblp.uni-trier.de/pid/168/8498.html), [Daniel Kroening](https://dblp.uni-trier.de/pid/k/DanielKroening.html), [Elizabeth Polgreen](https://dblp.uni-trier.de/pid/183/7353.html):
DSSynth: an automated digital controller synthesis tool for physical plants. [ASE 2017](https://dblp.uni-trier.de/db/conf/kbse/ase2017.html#AbateBCCCDKKP17): 919-924
[5] [Renato B. Abreu](https://dblp.uni-trier.de/pid/120/6549.html), [Mikhail Y. R. Gadelha](https://dblp.uni-trier.de/pid/127/1726.html), Lucas C. Cordeiro, [Eddie Batista de Lima Filho](https://dblp.uni-trier.de/pid/25/6797.html), [Waldir Sabino da Silva Jr.](https://dblp.uni-trier.de/pid/120/9096.html):
Bounded model checking for fixed-point digital filters. [J. Braz. Comput. Soc. 22(1)](https://dblp.uni-trier.de/db/journals/jbcs/jbcs22.html#AbreuGCFS16): 1:1-1:20 (2016)
[6] [Alessandro Trindade](https://dblp.uni-trier.de/pid/168/8729.html), Lucas C. Cordeiro:
Synthesis of Solar Photovoltaic Systems: Optimal Sizing Comparison. [VSTTE 2020](https://dblp.uni-trier.de/db/conf/vstte/vstte2020.html#TrindadeC20): 87-105
[7] [Rodrigo F. Araujo](https://dblp.uni-trier.de/pid/199/1650.html), [Higo F. Albuquerque](https://dblp.uni-trier.de/pid/199/2046.html), [Iury Valente de Bessa](https://dblp.uni-trier.de/pid/143/7442.html), Lucas C. Cordeiro, [João Edgar Chaves Filho](https://dblp.uni-trier.de/pid/120/9134.html):
Counterexample guided inductive optimization based on satisfiability modulo theories. [Sci. Comput. Program. 165](https://dblp.uni-trier.de/db/journals/scp/scp165.html#AraujoABCF18): 3-23 (2018)
[8] [Higo F. Albuquerque](https://dblp.uni-trier.de/pid/199/2046.html), [Rodrigo F. Araujo](https://dblp.uni-trier.de/pid/199/1650.html), [Iury Valente de Bessa](https://dblp.uni-trier.de/pid/143/7442.html), Lucas C. Cordeiro, [Eddie Batista de Lima Filho](https://dblp.uni-trier.de/pid/25/6797.html):
OptCE: A Counterexample-Guided Inductive Optimization Solver. [SBMF 2017](https://dblp.uni-trier.de/db/conf/sbmf/sbmf2017.html#AlbuquerqueABCF17): 125-141
[9] [Alessandro Bezerra Trindade](https://dblp.uni-trier.de/pid/168/8729.html), [Renato De Faria Degelo](https://dblp.uni-trier.de/pid/209/3497.html), [Edilson Galvão Dos Santos Junior](https://dblp.uni-trier.de/pid/209/4272.html), [Hussama Ibrahim Ismail](https://dblp.uni-trier.de/pid/171/6252.html), [Helder Cruz Da Silva](https://dblp.uni-trier.de/pid/209/3168.html), Lucas Carvalho Cordeiro:
Multi-core model checking and maximum satisfiability applied to hardware-software partitioning. [Int. J. Embed. Syst. 9(6)](https://dblp.uni-trier.de/db/journals/ijes/ijes9.html#TrindadeDJISC17): 570-582 (2017)

**Comment by fbrausse** [comment]
I'm running this over the SV-COMP benchmarks in <https://github.com/esbmc/esbmc/actions/runs/5532336687>. If they are OK, as discussed with @lucasccordeiro I'll remove `--ir` from some of the test cases that are failing now because of the change to `overflows_boundaries` (or at least mark them KNOWNBUG and add versions without the `--ir`). Besides that I'm still looking into the alignment pessimization, because removing that fixes all the failing cases. Any hints on how `object_descriptor2t::alignment` and the `deref_expr` are related would be appreciated.

File: src/pointer-analysis/dereference.cpp (hunk 1 above)
Comment (fbrausse) [reply src/pointer-analysis/dereference.cpp]: I've added comments explaining the loop and the (somewhat hidden) assumption this function makes.

**Comment by fbrausse** [comment]
stats-30s.txt:
```
Statistics:          26459 Files
  correct:           14414
    correct true:     8195
    correct false:    6219
  incorrect:            62
    incorrect true:     28
    incorrect false:    34
  unknown:           11983
  Score:             21169 (max: 43043)
```
[This PR](https://github.com/esbmc/esbmc/actions/runs/5532336687/jobs/10094955026) (a3281cf0b47c27df52ab7764ecfd40ba6fe9db88):
```
Statistics:          26459 Files
  correct:           14425
    correct true:     8204
    correct false:    6221
  incorrect:            59
    incorrect true:     28
    incorrect false:    31
  unknown:           11975
  Score:             21237 (max: 43043)
```
I'm running master again. Once it's finished I can start analyzing the differences.

**Comment by lucasccordeiro** [comment]
@fbrausse: Your PR already improves the master, right?

**Comment by fbrausse** [comment]
> @fbrausse: Your PR already improves the master, right?

Looks like, yes. :)

**Comment by fbrausse** [comment]
The incorrect ones are not fixed by this PR, they just became timeouts. Locally, they still produce the same verdict as master. In a debug build they fail in symex_va_arg(), see #584 for --64.

Btw. these are the master (18d016a377ffe9e73b1b4ef40c9cabd90a068d39) stats I [did get](https://github.com/esbmc/esbmc/actions/runs/5540520772) for 30s:
```
Statistics:          26459 Files
  correct:           14415
    correct true:     8194
    correct false:    6221
  incorrect:            61
    incorrect true:     28
    incorrect false:    33
  unknown:           11983
  Score:             21185 (max: 43043)
```

**Comment by fbrausse** [comment]
@lucasccordeiro I've reverted removing the `--ir` from the 3 test cases as they would fail without it, not only in this PR but also on master.

**Comment by lucasccordeiro** [comment]
Thanks, @fbrausse. 

Can I ask you to fix the conflicting file `.github/workflows/benchexec.yml`?

**Review by lucasccordeiro — APPROVED** [review]
LGTM.

**Comment by lucasccordeiro** [comment]
@fbrausse: can you please fix the conflicting file `.github/workflows/benchexec.yml`?

**Comment by fbrausse** [comment]
I'll add a (more proper) check for reducing the guaranteed alignment in `build_reference_to()` and would like to check this against SV-COMP once #1199 is merged in order to have more meaningful results. The last run did have noise from the intrinsic_memset errors.

**Comment by fbrausse** [comment]
> Some --fixedbv and some --ir regressions start to fail with this PR [...]

Not anymore, I found the bug: bitcasts from/to fixedbv were interpreted as typecasts in smt-conv and by the irep2 simplifier. They are not, though. I've added a regression test about that, which fails on master.

**Comment by fbrausse** [comment]
> I'll add a (more proper) check for reducing the guaranteed alignment in `build_reference_to()` and would like to check this against SV-COMP once #1199 is merged in order to have more meaningful results. The last run did have noise from the intrinsic_memset errors.

I've reverted my changes regarding the alignment guarantee in the dereference module. This needs more work and most likely it is the wrong place for this computation, too: should probably go into the value-set analysis. I'll run SV-COMP due to the other changes to dereference, though. Is 30s enough?

**Comment by fbrausse** [comment]
[Master](https://github.com/esbmc/esbmc/actions/runs/5620760754) (4640d2ec1d68849dc295d39c9d6c269f2ada1344):
```
Statistics:          26459 Files
  correct:           14419
    correct true:     8201
    correct false:    6218
  incorrect:            63
    incorrect true:     29
    incorrect false:    34
  unknown:           11977
  Score:             21148 (max: 43043)
```
[This PR](https://github.com/esbmc/esbmc/actions/runs/5643361092) (5f49e359b31e60d03a81120812e76583dc1a7141):
```
Statistics:          26459 Files
  correct:           14414
    correct true:     8196
    correct false:    6218
  incorrect:            59
    incorrect true:     29
    incorrect false:    30
  unknown:           11986
  Score:             21202 (max: 43043)
```
The 4 less incorrect ones are timeouts, they're not fixed by this PR. Otherwise there is some slow-down mainly in unreach-call. On the positive side, at least one got fixed:
- c/ntdrivers/kbfiltr.i.cil-1.c was before `dereference failure: Oversized field offset` (unknown), is now `dereference failure: Access to object out of bounds` (correct)

Nonetheless, I think this slow-down not necessary. It is due to my change determining `is_correctly_aligned` and `overflows_boundaries` which now looks like this for the dynamic offset case:
```c
    // Dyn offset -- is alignment guarantee strong enough?
    is_correctly_aligned = (alignment >= subtype_size);
    overflows_boundaries = !is_correctly_aligned || deref_size > subtype_size;
```
This can be fixed: The alignment guarantee passed to construct_from_array() is unnecessarily too pessimistic in many cases (including the regression test which timeouts atm.).

I'll give it another try in this PR, even though the place I'm putting the calculation of this particular quantity may not be the optimal choice...

**Comment by fbrausse** [comment]
> This can be fixed: The alignment guarantee passed to construct_from_array() is unnecessarily too pessimistic in many cases (including the regression test which timeouts atm.).
> 
> I'll give it another try in this PR, even though the place I'm putting the calculation of this particular quantity may not be the optimal choice...

I don't want to fix it in this PR: too many assumptions piling up, one of which is already known to be false, see #1220. We can decide whether to go ahead with this PR and leave the alignment pessimization untouched or whether to wait for a fix of #1220 first.

What do you think?

**Comment by lucasccordeiro** [comment]
@fbrausse: I would recommend moving forward with this PR and leaving the alignment passivization untouched for another PR.

**Comment by lucasccordeiro** [comment]
@fbrausse: can you please fix the conflicting file of this PR?

**Comment by fbrausse** [comment]
@lucasccordeiro I have a few patches ontop of this branch which are not pushed, yet. One of them says "WIP: cannot actually avoid byte-extracting from arrays". Might be that another solution is required, I don't quite remember in detail why I thought that is case, this PR has been hanging around for too long. I'll mark it a draft until I found out.

**Comment by fbrausse** [comment]
Alright, everything should be sorted out, I think.

Quick recap because it's been a long time since this PR was first described. It does:
1. consolidate the extract_bytes_*() functions into a unified one handling arbitrary levels of arrays
2. strengthen the conditions determining overflows_boundaries for extracting elements from arrays, and
3. remove an unnecessary bitcast-to-BV before byte-extraction in smt_conv + some fixes to fixedbv conversion.

Items 1 and 2 fix #169: it's not just Boolector but basically all solvers, the problem was about byte-extract from VLAs.

As discussed in #1215, item 2 above affects a few `--ir --fixedbv` benchmarks, basically because the deref code doesn't trust the alignment info from the value-set when the `deref_expr` isn't a simple symbol. As discussed, this PR is not touching that, because (a) the assumption (is_symbol2t) is wrong, see #1220, and (b) fixing it conservatively has a too high performance impact.

I believe it's ready now.

PS: I'm running the benchmarks mainly because of items 1 and 2. Don't expect fixedbv issues, but there could be ones concerning non-byte-arrays.

**Review by lucasccordeiro — APPROVED** [review]

**Review by mikhailramalho — APPROVED** [review]
lgtm

**Comment by fbrausse** [comment]
From stats-30s.txt on [master](https://github.com/esbmc/esbmc/actions/runs/6298655982) (6c7905dc0851524bac34f14687b8b6491906dd9b) (do we have more recent numbers?):
```
Statistics:          23805 Files
  correct:           14173
    correct true:     7824
    correct false:    6349
  incorrect:            22
    incorrect true:      7
    incorrect false:    15
  unknown:            9610
  Score:             21533 (max: 38482)
```
[This PR](https://github.com/esbmc/esbmc/actions/runs/6416568906) (7756482b92244f6f68155c2ff0177832df9bb412):
```
Statistics:          23805 Files
  correct:           14161
    correct true:     7817
    correct false:    6344
  incorrect:            21
    incorrect true:      6
    incorrect false:    15
  unknown:            9623
  Score:             21546 (max: 38482)
```
There are 3x timeout -> unknown in unreach-call: wrong assertion triggered by Linux driver benchmarks here for a constant_string value: https://github.com/fbrausse/esbmc/blob/73ef67cc75b4e7d8bccd0a5466e3c60afd99c0c9/src/pointer-analysis/dereference.cpp#L1551 This PR didn't touch that function. It is most likely caused by not byte-extracting from constant_strings anymore, but by indexing into them instead making them part of the value-set.

The 1 incorrect -> timeout is
- c/pthread/triangular-longest-2.i in unreach-call

It's still incorrect with this PR, just did take a bit longer.

**Comment by lucasccordeiro** [comment]
Thanks for submitting this PR, @fbrausse.

**Pull request merged**